### PR TITLE
(MISC) Mark Rust tokens as operators

### LINKED ISSUE #668: Support operator highlighting
`intellij-rust` currently has a Colors & Fonts setting for "Operator Sign", but it does not seem to be functional.

For reference:
- [Unary operator expressions](https://doc.rust-lang.org/reference.html#unary-operator-expressions)
- [Binary operator expressions](https://doc.rust-lang.org/reference.html#binary-operator-expressions)

[highlighterDemoText.rs](https://github.com/intellij-rust/intellij-rust/blob/9403901249c1db15ed8cdc49a48554d54bdd3666/src/main/resources/org/rust/ide/colors/highlighterDemoText.rs#L35) doesn't have `+=` marked up as an operator, for example.

Tested in
- `intellij-rust` 0.1.0.981
- PyCharm Pro 2016.2.3

### PR DESCRIPTION
**Pull request by allengeorge**
Closes #668

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt [hunk 1]
```diff
@@ -47,6 +47,24 @@ class RustHighlighter : SyntaxHighlighterBase() {
             LBRACE, RBRACE                 -> RustColor.BRACES
             LBRACK, RBRACK                 -> RustColor.BRACKETS
 
+            UNDERSCORE, DOLLAR             -> RustColor.OPERATORS
```
Comment (allengeorge) [inline src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt]: Not sure about the first three.

File: src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt [hunk 2]
```diff
@@ -47,6 +47,24 @@ class RustHighlighter : SyntaxHighlighterBase() {
...
+            XOR, XOREQ                     -> RustColor.OPERATORS
+            EXCL, EXCLEQ                   -> RustColor.OPERATORS
+            MUL, MULEQ                     -> RustColor.OPERATORS
+            DIV, DIVEQ                     -> RustColor.OPERATORS
+            REM, REMEQ                     -> RustColor.OPERATORS
+            EQ, EQEQ                       -> RustColor.OPERATORS
+            PLUS, PLUSEQ                   -> RustColor.OPERATORS
+            MINUS, MINUSEQ                 -> RustColor.OPERATORS
+            GT, LT                         -> RustColor.OPERATORS
+            Q, AT                          -> RustColor.OPERATORS
```
Comment (matklad) [inline src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt]: I think most of this is in this token set: https://github.com/intellij-rust/intellij-rust/blob/master/src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java#L174

Perhaps it is worth making another token set for other operators.

Comment (mkaput) [inline src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt]: I think rolling out `RustOperatorTokenType` ([`RustTokenType.kt`](https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/lang/core/psi/RustTokenType.kt)) and then:
```kotlin
val OPERATORS = TokenSet.create(*IElementType.enumetare({ it is RustOperatorTokenType })
```
may be good idea.

File: src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt (hunk 1 above)
Comment (mkaput) [inline src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt]: '$' and '#' are fine, underscore should be rather treated as identifier imho

File: src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt (hunk 2 above)
Comment (allengeorge) [inline src/main/kotlin/org/rust/ide/highlight/RustHighlighter.kt]: Thanks guys! - I'll fix this up and resubmit.

File: src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java [hunk 3]
```diff
@@ -69,6 +69,7 @@
 
     RustTokenType IDENTIFIER = new RustTokenType("<IDENTIFIER>");
     RustTokenType LIFETIME = new RustTokenType("<LIFETIME>");
+    RustTokenType UNDERSCORE = new RustTokenType("_");
```
Comment (allengeorge) [inline src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java]: Moved this up out of the operators section

File: src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java [hunk 4]
```diff
@@ -203,6 +203,18 @@
         ANDAND
     );
 
+    @NotNull
+    TokenSet ALL_OPS = TokenSet.create(
```
Comment (allengeorge) [inline src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java]: Not sure if I actually need this token set; I don't use it at all. I guess it'll be useful elsewhere?

Comment (mkaput) [inline src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java]: Hmm, [SPACE_AROUND_OPS](https://github.com/intellij-rust/intellij-rust/blob/master/src/main/kotlin/org/rust/ide/formatter/impl/utils.kt#L17) may be redefined as `ALL_OPS \ NO_SPACE_AROUND_OPS`

**Comment by mkaput** [comment]
> Also, FWIW, I get a build error on master?

No, you're waiting for @matklad's review approval

**Comment by allengeorge** [comment]
@mkaput No - I mean that I run `gradlew build` locally on master and I get one test failure.

File: src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java (hunk 4 above)
Comment (allengeorge) [inline src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java]: Whoa. There's a `\` operator?

**Comment by mkaput** [comment]
Which test does fail on your machine? What's the error message?

File: src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java (hunk 4 above)
Comment (mkaput) [inline src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java]: Nope, I used mathematical operator for set complement (substracting) :P

Comment (mkaput) [inline src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java]: In IntelliJ's TokenSet language: `SPACE_AROUND_OPS = TokenSet.andNot(ALL_OPS, NO_SPACE_AROUND_OPS);`

**Comment by allengeorge** [comment]
`RunConfigurationTestCase.testApplicationConfiguration`

```
java.lang.RuntimeException: java.lang.AssertionError: 
Expecting:
 <"">
to contain:
 <"Hello, world!"> 
	at com.intellij.openapi.application.impl.LaterInvocator.invokeAndWait(LaterInvocator.java:177)
	at com.intellij.openapi.application.impl.ApplicationImpl.invokeAndWait(ApplicationImpl.java:642)
	at com.intellij.testFramework.EdtTestUtilKt.runInEdtAndWait(EdtTestUtil.kt:42)
	at com.intellij.testFramework.EdtTestUtil$Companion.runInEdtAndWait(EdtTestUtil.kt:29)
	at com.intellij.testFramework.EdtTestUtil.runInEdtAndWait(EdtTestUtil.kt)
	at com.intellij.testFramework.PlatformTestCase.runBareRunnable(PlatformTestCase.java:717)
	at com.intellij.testFramework.PlatformTestCase.runBareImpl(PlatformTestCase.java:692)
	at com.intellij.testFramework.PlatformTestCase.runBare(PlatformTestCase.java:628)
	at junit.framework.TestResult$1.protect(TestResult.java:122)
	at junit.framework.TestResult.runProtected(TestResult.java:142)
	at junit.framework.TestResult.run(TestResult.java:125)
	at junit.framework.TestCase.run(TestCase.java:129)
	at junit.framework.TestSuite.runTest(TestSuite.java:252)
	at junit.framework.TestSuite.run(TestSuite.java:247)
	at org.junit.internal.runners.JUnit38ClassRunner.run(JUnit38ClassRunner.java:86)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.runTestClass(JUnitTestClassExecuter.java:114)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecuter.execute(JUnitTestClassExecuter.java:57)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassProcessor.processTestClass(JUnitTestClassProcessor.java:66)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:109)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:377)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:54)
	at org.gradle.internal.concurrent.StoppableExecutorImpl$1.run(StoppableExecutorImpl.java:40)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.AssertionError: 
Expecting:
 <"">
to contain:
 <"Hello, world!"> 
	at org.rust.cargo.runconfig.RunConfigurationTestCase$testApplicationConfiguration$1.invoke(RunConfigurationTestCase.kt:20)
	at org.rust.cargo.runconfig.RunConfigurationTestCase$testApplicationConfiguration$1.invoke(RunConfigurationTestCase.kt:13)
	at org.rust.cargo.RustWithToolchainTestBase.withProject(RustWithToolchainTestBase.kt:27)
	at org.rust.cargo.runconfig.RunConfigurationTestCase.testApplicationConfiguration(RunConfigurationTestCase.kt:16)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at junit.framework.TestCase.runTest(TestCase.java:176)
	at com.intellij.testFramework.UsefulTestCase.access$001(UsefulTestCase.java:80)
	at com.intellij.testFramework.UsefulTestCase.lambda$runTest$0(UsefulTestCase.java:326)
	at com.intellij.testFramework.UsefulTestCase$$Lambda$146/488031859.run(Unknown Source)
	at com.intellij.testFramework.PlatformTestCase.lambda$invokeTestRunnable$3(PlatformTestCase.java:737)
	at com.intellij.testFramework.PlatformTestCase$$Lambda$147/945486519.run(Unknown Source)
	at com.intellij.testFramework.PlatformTestCase.invokeTestRunnable(PlatformTestCase.java:749)
	at com.intellij.testFramework.UsefulTestCase.runTest(UsefulTestCase.java:341)
	at org.rust.cargo.RustWithToolchainTestBase.runTest(RustWithToolchainTestBase.kt:35)
	at com.intellij.testFramework.PlatformTestCase.lambda$runBareImpl$2(PlatformTestCase.java:666)
	at com.intellij.testFramework.PlatformTestCase$$Lambda$11/1241860672.run(Unknown Source)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:29)
	at com.intellij.testFramework.EdtTestUtil$Companion$runInEdtAndWait$1.invoke(EdtTestUtil.kt:27)
	at com.intellij.testFramework.EdtTestUtilKt$sam$Runnable$ca4818df.run(EdtTestUtil.kt)
	at com.intellij.openapi.application.TransactionGuardImpl$6.run(TransactionGuardImpl.java:291)
	at com.intellij.openapi.application.impl.LaterInvocator$1.run(LaterInvocator.java:156)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:326)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:310)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:749)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:702)
	at java.awt.EventQueue$3.run(EventQueue.java:696)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$1.doIntersectionPrivilege(ProtectionDomain.java:75)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:719)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:863)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:699)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:385)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by matklad** [comment]
That tests literally executes `cargo run` using the Cargo from `PATH`, so it may fail if there's something unusual about your rust setup.

**Comment by allengeorge** [comment]
@matklad Ah. I see. I'll poke around and see what's up. Thanks for the clarification. I was really puzzled because I couldn't figure out why my change would have triggered that error.

File: src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java [hunk 5]
```diff
@@ -95,61 +96,60 @@
...
+    RustTokenType MULEQ = new RustOperatorTokenType("*=");
+    RustTokenType OR = new RustOperatorTokenType("|");
+    RustTokenType OREQ = new RustOperatorTokenType("|=");
+    RustTokenType PLUS = new RustOperatorTokenType("+");
+    RustTokenType PLUSEQ = new RustOperatorTokenType("+=");
+    RustTokenType RBRACE = new RustOperatorTokenType("}");
+    RustTokenType RBRACK = new RustOperatorTokenType("]");
+    RustTokenType REM = new RustOperatorTokenType("%");
+    RustTokenType REMEQ = new RustOperatorTokenType("%=");
+    RustTokenType RPAREN = new RustOperatorTokenType(")");
```
Comment (matklad) [inline src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java]: I think that `(`, `{`, `[` braces should not be colored as operators (they are not in Kotlin).

Comment (allengeorge) [inline src/main/java/org/rust/lang/core/psi/RustTokenElementTypes.java]: Good point. Fixed, and put it under a separate block of tokens named "Grouping"

**Comment by allengeorge** [comment]
@mkaput @matklad Thank you so much for bearing with me and helping me get this PR in!

**Comment by matklad** [comment]
Thanks @allengeorge !

**Pull request merged**